docs/proposals/http-tunnel-expose-strategy.md

```diff
@@ -0,0 +1,137 @@
+
+
+### Accessing the Kubelet from the Apiserver through the Openvpn tunnel
+
+Apiserver needs to communicate with the Kubelet on the user cluster nodes when using commands like “kubectl logs” or “kubectl exec”. Routing of traffic follows the same steps as the previous section. Openvpn will send the traffic to the HTTP Tunneler. This will get forwarded to the HTTP Proxy that routes it, thanks to the authority header, to the final Openvpn-server ClusterIP service.
+
+
+<div style="text-align:center">
+  <img src="images/http-tunnel-expose-strategy-full-tunneling.png" width="100%" />
+</div>
+
+### Setup involving a corporate HTTP Proxy
+
+In case worker nodes do not have direct internet access but have to pass
+through a corporate HTTP proxy (we will call it middle HTTP proxy in the rest
+of the document), we need to configure the HTTP tunneler to send traffic to the
+middle HTTP proxy first.
+
+There are two ways of achieving this:
+
+* Configuring the middle HTTP proxy to forward the connect requests addressed
+    to the seed HTTP proxy. This strategy is not transparent to the middle HTTP
+    proxy and requires some specific configuration.
+* Having double CONNECT encapsulation like depicted below:
+    ```
+    Pod-->[HTTPS]-->HTTP Tunneler--[HTTPS tunneled over HTTP/2 tunneled over HTTP/1.1]-->Middle HTTP Proxy-->[HTTPS tunneled over HTTP/2]-->Seed HTTP Proxy--[HTTPS]-->Kube Apiserver/Openvpn
+    ```
+    The advantage of this solution is that it does not need specific
+    configuration in the middle HTTP proxy.
+
+### Security considerations
+
+> A generic TCP tunnel is fraught with security risks. First, such
+> authorization should be limited to a small number of known ports.
+> The Upgrade: mechanism defined here only requires onward tunneling at
+> port 80. Second, since tunneled data is opaque to the proxy, there
+> are additional risks to tunneling to other well-known or reserved
+> ports. A putative HTTP client CONNECTing to port 25 could relay spam
+> via SMTP, for example.
+> -- <cite>[rfc2817][http_upgrade_tls_rfc]</cite>
+
+The HTTP Proxy on seed cluster must only accept CONNECT requests addressed to
+the OpenVPN server and Kube Apiserver, all other requests should be dropped.
+
+Moreover and to prevent disclosure of information sent in cleartext between the 
+CONNECT requests, we have implemented mTLS between the two Envoy pods 
+(HTTP Tunneler on the user cluster and HTTP Proxy on the seed cluster). 
+
+## **Alternatives considered**
+
+*   Using SNI everywhere: in this case, we would use the OpenVPN3 client instead of OpenVPN2 client. This one supports SNI. The problem we encounter is that OpenVPN uses the OpenVPN protocol which is based on TLS. In OpenVPN protocol the ClientHello is not the first packet exchanged after the TCP handshake, thus standard SNI routers cannot be used. This behaviour has been tested with Nginx ingress controller and Envoy.
```

Comment:
We should evaluate https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/ for k8s >= 1.18 as a drop-in replacement for OpenVPN